app/src/main/java/tttTDD/Java/Board.java

```diff
@@ -1,44 +1,90 @@
 
 public class Board {
+    private final int DEFAULT_SIZE = 3;
     private final String[] board;
+    private int numberOfSpaces;
+    private int[][] winningSets = {
+            {0, 1, 2},
+            {3, 4, 5},
+            {6, 7, 8},
+            {0, 3, 6},
+            {1, 4, 7},
+            {2, 5, 8},
+            {0, 4, 8},
+            {2, 4, 6},
+    };
 
     public Board() {
-        this.board = new String[] {"1", "2", "3", "4", "5", "6", "7", "8", "9"};
+        this.board = initializeBoard(DEFAULT_SIZE);
     }
 
     public Board(int boardSize) {
-        this.board = new String[boardSize * boardSize];
-        for (int i = 0; i < this.board.length; i++) {
-            this.board[i] = String.valueOf(i + 1);
-        }
+        this.board = initializeBoard(boardSize);
     }
 
-    public String[] getBoard() {
-        return this.board;
+    public String getSpace(int location) {
+        return isLocationValid(location) ? null : this.board[location];
     }
 
-    public String getSpace(int location) {
-        return this.board[location];
+    public int getFirstAvailableMove() {
+        String[] availableMoves = getAvailableMoves(board);
+        return convertToLocation(availableMoves[0]);
     }
 
-    public void updateBoard(int location, String marker) throws Exception {
+    public boolean updateBoard(int location, String marker) {
         if (isSpaceAvailable(location)) {
             this.board[location] = marker;
-        } else {
-            throw new Exception("The space is already occupied");
+            return true;
+        }
+
+        return false;
+    }
+
+    public boolean hasWinningSet() {
+        for (int[] set : winningSets) {
+            if (board[set[0]].equals(board[set[1]]) && board[set[1]].equals(board[set[2]])) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String[] initializeBoard(int boardSize) {
+        numberOfSpaces = boardSize * boardSize;
+        String[] board = new String[numberOfSpaces];
+        for (int i = 0; i < board.length; i++) {
+            board[i] = String.valueOf(i + 1);
         }
+
+        return board;
     }
 
     private boolean isSpaceAvailable(int location) {
         return (String.valueOf(location + 1).equals(getSpace(location)));
     }
+
+    private int convertToLocation(String space) {
+        return Integer.parseInt(space) - 1;
+    }
+
+    private boolean isLocationValid(int location) {
+        return (location < 0) || (location > numberOfSpaces - 1);
+    }
+
+    private String[] getAvailableMoves(String[] board) {
+        ArrayList<String> availableMovesArray = new ArrayList<>();
+
+        for (int i = 0; i < board.length; i++) {
+            if (isSpaceAvailable(i)) {
+                availableMovesArray.add(board[i]);
+            }
+        }
+
+        return Arrays.copyOf(availableMovesArray.toArray(), availableMovesArray.size(), String[].class);
+    }
```

Comment:
This is my opinion but I think it would be simpler/cleaner to use a stream here

```
return Arrays.stream(board).filter(this::isSpaceAvailable);
```

I believe this is functionally equivalent to what you're doing here

Reply:
I tried using the snippet up top and tried modifying it so that isSpace receives the location parameter but it didn't play nice with it.
`Arrays.stream(board).filter(space -> isSpaceAvailable(Integer.parseInt(space) - 1))`

it was asking for the function to return `Stream<String>` when `String[]` is needed.

Reply:
per https://github.com/rjdeloss/tictactoe-v2-tdd-java/pull/3#commitcomment-73948265 this can be left for another time